kafka/src/test/groovy/io/micronaut/configuration/kafka/annotation/KafkaClientWithExecutorSpec.groovy

```diff
@@ -0,0 +1,221 @@
+package io.micronaut.configuration.kafka.annotation
+
+import io.micronaut.configuration.kafka.AbstractKafkaSpec
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.annotation.Requires
+import io.micronaut.core.io.socket.SocketUtils
+import io.micronaut.scheduling.TaskExecutors
+import org.apache.kafka.common.header.Headers
+import org.apache.kafka.common.header.internals.RecordHeader
+import org.apache.kafka.common.header.internals.RecordHeaders
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import reactor.core.publisher.Mono
+import spock.lang.AutoCleanup
+import spock.util.concurrent.PollingConditions
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicInteger
+
+import static io.micronaut.configuration.kafka.annotation.KafkaClient.Acknowledge.ALL
+import static io.micronaut.core.io.socket.SocketUtils.LOCALHOST
+
+class KafkaClientWithExecutorSpec extends AbstractKafkaSpec {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaClientWithExecutorSpec.class)
+
+    @AutoCleanup
+    private ApplicationContext ctx
+
+    void setup() {
+        ctx = ApplicationContext.run(
+                getConfiguration() +
+                        ['kafka.bootstrap.servers': LOCALHOST + ':' + SocketUtils.findAvailableTcpPort(),
+                         'kafka.producers.default.executor': TaskExecutors.BLOCKING,
+                         'kafka.producers.scorsese.executor': TaskExecutors.IO])
+
+    }
+
+    void "test reactive send message with executor set via default config props  does not block the calling thread when Kafka is not available"() {
+        given:
+        MyDefaultExecutorClient client = ctx.getBean(MyDefaultExecutorClient)
+        PollingConditions conditions = new PollingConditions()
+        conditions.initialDelay = 1
+
+        when:
+        AtomicInteger failureCount = new AtomicInteger(0)
+        List<Mono<String>> sendOps = new ArrayList<>()
+        for (int x=0; x<3; x++) {
+            sendOps.add(client.sendRx("test", "hello-world", new RecordHeaders([new RecordHeader("hello", "world".bytes)])))
+        }
+        sendOps.stream().map(sendOp -> {
+            sendOp.subscribe(s -> { throw new IllegalStateException("Unexpected result") },
+                    ex -> {
+                        int currentCount = failureCount.incrementAndGet()
+                        LOG.debug("Subscriber failure # {} : {}", currentCount, ex.getMessage())
+                    })
+        }).toList()
+
+        then:
+        conditions.eventually {
+            failureCount.get() < 3
+        }
```

Comment:
Waiting for things to be less than a number (when they start less than a number) make me nervous ;-)

Are we testing that it gets here before the messages are attempted?

If so, do we need the `eventually`?

And is there a way of testing the opposite?

I worry this test could be green, but the underlying behaviour is not doing what we think it is 🤔 

I may be wrong 😄 
If nothing has happened yet, it will be less than 3

Reply:
> Are we testing that it gets here before the messages are attempted?

Yes, basically. It's a test to prove that the operation is happening asynchronously rather than blocking the calling thread (which in this case is just the main test thread). If you remove the `executor` configuration, then the operations will block (up to the `maxBlock` specified in the `@KafkaClient` config) and this count will always be 3. The way it is written write now, I could probably safely say `failureCount.get() == 0` because of the timings involved (see below).

> If so, do we need the `eventually`?

We could conceivably get rid of the `eventually`. I was mostly using it to play with timings and ensure that the test actually fails if you change the `initialDelay` to be longer than the value of `maxBlock` in the `@KafkaClient` configurations.

> And is there a way of testing the opposite?

Yes, see comments above re: `executor` and `maxBlock`.

> I worry this test could be green, but the underlying behaviour is not doing what we think it is 🤔

> I may be wrong 😄 If nothing has happened yet, it will be less than 3

I think I can simplify - get rid of the `eventually` and make the test `failureCount.get() == 0`, but I'm concerned that might not work consistently in all environments due to differences in thread scheduling. I'll give it a try. If in that state it works locally (and I can still break it by removing the `executor` config), works in CI, and works on someone's ARM processor, then it's probably safe enough.

I will also add some more descriptive text to the steps of the spec methods since the intent is not totally obvious.

Reply:
Ok, I've cleaned it up in [c9b2b93](https://github.com/micronaut-projects/micronaut-kafka/pull/955/commits/c9b2b93a3bf73302f98b66430a44239ad4e87e3d)

I also made the `ApplicationContext` configuration specific to each spec method - otherwise just removing `executor` from one place where it's specified via `@KafkaClient` annotation doesn't cause those methods to fail as expected because it was falling back to the default configuration (which is appropriate, but was the not the condition that was meant to be tested).

Reply:
I *could* add an additional `eventually` condition to assert that the failures do eventually happen as expected - but that would make the test slower to run which I always prefer to avoid.